### PR TITLE
Dana's dev blog

### PR DESCRIPTION
**Pull request by vsaulue**
### Author info
Name: Credne
Release date: 12th August 2022

#### Requirements for media in submissions:

- [x] Images are hosted on Media server 
- [x] Upload videos

### CODE REVIEW

File: submissions/dana-channel-routing/main.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+Dana is born out of frustration of having to apply this process for hours on complex mod overhauls, and it takes a *slightly* different approach to answer these questions:
+
+![dana-demo](dana-demo.mp4)
+("How to make *Chemical science packs*?")
+
+Dana is about depth: When asked how to make a Chemical science pack, it will show you all the chained steps required from the raw materials. And it does so directly in-game, by drawing you a *nice™ and understandable™* recipe graph. Players can navigate the graph with WASD, zoom in/out with mouse wheel, and select nodes or links for additional info. It's also possible to draw the full crafting graph of the current game (the vanilla one will be shown further in this article), or an "usage" graph (ie. what are all the items that can be crafted from *X*).
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
Dana is about depth: When asked how to make a Chemical science pack, it will show you all the chained steps required from the raw materials. And it does so directly in-game, by drawing you a *nice™ and understandable™* recipe graph. Players can navigate the graph with WASD, zoom in/out with mouse wheel, and select nodes or links for additional info. It's also possible to draw the full crafting graph of the current game (the vanilla one will be shown further in this article), or a "usage" graph (i.e., what are all the items that can be crafted from *X*).
```

File: submissions/dana-channel-routing/main.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+Dana is designed to work out of the box with any combination of mods adding/modifying/removing recipes or items. There is no hardcoded configuration provided by the player, the modders, or shipped directly with Dana. In this video, it placed the "Copper block" between the "Iron block" and the "Refinery block", the "Heavy oil" above the "Light oil", decided how the lines should be drawn/merged/bended, what each block's X-Y coordinates are, and more. All that Dana had to work with was the full list of items, recipes, and what are the available natural resources. This is done by a *graph layout algorithm* (the piece of code that takes a bunch of recipes/items and decides where to place them on screen and how to link them) specifically designed for Factorio, which will be the subject of this article.
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
Dana is designed to work out of the box with any combination of mods adding/modifying/removing recipes or items. There is no hardcoded configuration provided by the player, the modders, or shipped directly with Dana. In this video, it placed the "Copper block" between the "Iron block" and the "Refinery block", the "Heavy oil" above the "Light oil", decided how the lines should be drawn/merged/bended, what each block's X-Y coordinates were, and more. All that Dana had to work with was the full list of items, recipes, and what the available natural resources were. This is done by a *graph layout algorithm* (the piece of code that takes a bunch of recipes/items and decides where to place them on the screen and how to link them) specifically designed for Factorio, which will be the subject of this article.
```

File: submissions/dana-channel-routing/main.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+### Dev Blog: Drawing the Graph's Spaghetti
+
+Today's post will present some (hopefully interesting) details about the inner working of Dana's *nice™ and understandable™* graph generator. To give a general introduction, Dana's graphs are a variant of [layered graphs](https://en.wikipedia.org/wiki/Layered_graph_drawing). This means that items and recipes are placed in *node layers*, separated by *link layers*.
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
Today's post will present some (hopefully interesting) details about the inner workings of Dana's *nice™ and understandable™* graph generator. To give a general introduction, Dana's graphs are a variant of [layered graphs](https://en.wikipedia.org/wiki/Layered_graph_drawing). This means that items and recipes are placed in *node layers*, separated by *link layers*.
```

File: submissions/dana-channel-routing/main.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+![Layered graph](layers-illustration.png)
+(Layered graph structure: node layers  with the blue background, link layers with the green background.)
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
(Layered graph structure: node layers with the blue background, link layers with the green background.)
```

File: submissions/dana-channel-routing/main.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+The first thing Dana does is decide how many node layers are needed, and in which layer each item/recipe will be placed. The second step is to decide the horizontal coordinate of each item/recipe. The third step is building the link layers. The final step is to assign a vertical coordinate to each element in the graph, now that the number of layers and their height is known.
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
The first thing Dana does is to decide how many node layers are needed, and in which layer each item/recipe will be placed. The second step is to decide the horizontal coordinate of each item/recipe. The third step is to build the link layers. The final step is to assign a vertical coordinate to each element in the graph, now that the number of layers and their height is known.
```

File: submissions/dana-channel-routing/main.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+The full layout algorithm is way too big and technical for an Alt-F4 article, so the rest of this will focus on the third step. Here's the problem: given two consecutive *node layers*, trace the required links in the *link layer*, in order to get a *nice™ and understandable™* graph:
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
The full layout algorithm is way too big and technical for an Alt-F4 article, so the rest of this will focus on the third step, which is building the link layers. Here's the problem: given two consecutive *node layers*, trace the required links in the *link layer*, in order to get a *nice™ and understandable™* graph:
```

File: submissions/dana-channel-routing/main.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+![Intro: problem description](intro-problem-description.png)
+(Input data of the problem)
+![Intro: possible solution](intro-problem-solution.png)
+(Possible result)
+
+Conveniently, this is more or less a variant of kindergarten exercises:
+
+![Kindergarten version](problem-kindergarten-version.png)
+
+Since five years old kids do that, it shouldn't be hard to program, right?
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
Since 5-year-olds do that, it shouldn't be hard to program, right?
```

File: submissions/dana-channel-routing/main.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+* Minimise line crossings, especially between almost parallel lines.
+* Minimise the bends on a link.
+* Minimise the length of links.
+
+On top of that, there are general UI design rules:
+
+* The user experience plummets if they have to scroll horizontally/vertically to cross-check information from different parts of the interface. To avoid that, the graphs must be as compact as possible.
+* Less is more: if the same amount of information can be conveyed with four lines instead of twenty, that's probably a good thing to do.
+
+So now, it's time to *search* for better ideas. And the best way to research for anything, as we all know, is ~~google~~ by pressing "T" in Factorio:
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
So now, it's time to *search* for better ideas. And the best way to research for anything, as we all know, is by ~~googling~~ pressing "T" in Factorio:
```

File: submissions/dana-channel-routing/main.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+![Factorio tech tree](factorio-tech-tree.png)
+
+Factorio's graph renderer has a more subtle way of rendering links. Each link is decomposed into three segments: two verticals, one horizontal. This approach scales much better with wider graphs, because:
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
Factorio's graph renderer has a more subtle way of rendering links. Each link is decomposed into three segments: two vertical, one horizontal. This approach scales much better with wider graphs, because:
```

File: submissions/dana-channel-routing/main.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+* There's never a crossing of nearly parallel lines. All crossings are at right angle, which is optimal to prevent users from following the wrong line.
+* The density of links is kept under control: there's always enough gap between parallel lines to distinguish them from each other.
+
+The price for this readability is vertical space: there need to be enough room between the two rows of technologies to add all the horizontal segments without collisions:
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
The price for this readability is vertical space: there needs to be enough room between two rows of technologies to be able to add all the horizontal segments without getting any collisions:
```

File: submissions/dana-channel-routing/main.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+
+![Factorio tech tree](factorio-tech-tree-spacing.png)
+
+To minimize this cost, there's a simple yet huge optimisation: What if lines didn't link just two elements, but any number of elements? Just draw a single wide horizontal line for each item/technology, then add as many vertical lines as necessary to connect to the nodes.
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
To minimize this cost, there's a simple yet huge optimisation: What if lines didn't link just two elements, but any number of elements? Just draw a single, wide horizontal line for each item/technology, and add as many vertical lines as necessary to connect to the nodes.
```

File: submissions/dana-channel-routing/main.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+![Dana: Factorio full graph](dana-full-graph.png)
+
+### The Coding Part
+
+So that's all for the drawing board, time to throw some code at the problem to get the job done! But with that come some other problems. First, Factorio Mods are made in a language called Lua, and Lua has an ridiculously barren ecosystem. No success in finding a library that can do this kind of link routing.
+
+Another solution would be porting a library from other languages. Sadly there are plenty of libraries to draw conventional graphs, but Dana is now dealing with links between any number of nodes. That's not a graph anymore, but an hypergraph. While the word sounds definitely cooler, there aren't many software libraries to draw them, and in general there's much less scientific literature on the topic. No success in finding clues to do routing the Dana way.
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
Another solution would be porting a library from other languages. Sadly there are plenty of libraries to draw conventional graphs, but Dana is now dealing with links between any number of nodes. That's not a graph anymore, but a hypergraph. While the word definitely sounds cooler, there aren't many software libraries to draw them, and in general there's much less scientific literature on the topic. No success in finding clues to do routing the Dana way.
```

File: submissions/dana-channel-routing/main.md [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+So, Dana has a router made "almost" from scratch. "Almost" because there was a lot of inspiration to be found elsewhere, it just required looking into some unexpected places...
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
So, Dana has a router made "almost" from scratch. "Almost", because there was a lot of inspiration to be found elsewhere, it just required looking into some unexpected places...
```

File: submissions/dana-channel-routing/main.md [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+
+#### Inspiration from PCB Design
+
+It happens that there are some people whose daily job requires to link points on a 2D plane: printed circuit board (PCB) designers. And for problems almost identical to Dana's link routing, they have a decade-old family of well documented algorithms: [channel routers](https://en.wikipedia.org/wiki/Channel_router).
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
There happen to be some people whose daily job requires linking points on a 2D plane: printed circuit board (PCB) designers. And for problems almost identical to Dana's link routing, they have a decade-old family of well-documented algorithms: [channel routers](https://en.wikipedia.org/wiki/Channel_router).
```

File: submissions/dana-channel-routing/main.md [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+![Picture of two channels](https://upload.wikimedia.org/wikipedia/commons/thumb/5/50/ChannelRouteSolution.png/800px-ChannelRouteSolution.png)
+
+Before looking at the solution, the first things Dana got from that is a proper way to modelize the problem. The goal of our link router is twofold: Determine the number of *channels* between the rows of nodes, and assign a *channel* to each horizontal segment of the links.
+
+Where each horizontal line starts and ends is simply determined by which nodes they must be linked to, and the vertical segments are simple projections from the nodes to the horizontal lines.
+
+![Channels and trunks](dana-channels-and-trunks.png)
+(Here, the router decided to make six channels in cyan, then choose a channel for each red horizontal segments.)
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
(Here, the router decided to make six channels in cyan, then choosing a channel for each red, horizontal segment.)
```

File: submissions/dana-channel-routing/main.md [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+The fundamental problem is the same. Luckily for us, the sports version is as old as round-robin tournaments, and the good thing about old mainstream problems is that there are a lot of smart people who have done research on it!
+
+A generic way to solve the issue is to use graph theory, where our sports problem would be equivalent to the [Minimum feedback arc set](https://en.wikipedia.org/wiki/Feedback_arc_set) problem. The bad news: it's a [NP-hard](https://en.wikipedia.org/wiki/NP-hardness) optimisation problem. In layman's terms: finding the best solution can be **ridiculously** time consuming even with just a few dozen of players. The good news: there is a nice pile of research articles proposing *heuristics*. Those algorithms' solutions might not be optimal, but "close enough" to optimal in "good enough" time. Various heuristics exist depending on how much computation time one is willing to pay in order to get stronger guarantees of optimality, or which can be tailored to specific types of graphs.
```
Comment (MyNameIsTrez) [inline submissions/dana-channel-routing/main.md]:
```suggestion
A generic way to solve the issue is to use graph theory, where our sports problem would be equivalent to the [Minimum feedback arc set](https://en.wikipedia.org/wiki/Feedback_arc_set) problem. The bad news: it's an [NP-hard](https://en.wikipedia.org/wiki/NP-hardness) optimisation problem. In layman's terms: finding the best solution can be **ridiculously** time consuming even with just a few dozen players. The good news: there is a nice pile of research articles proposing *heuristics*. Those algorithms' solutions might not be optimal, but "close enough" to optimal in "good enough" time. Various heuristics exist depending on how much computation time one is willing to pay in order to get stronger guarantees of optimality, or which can be tailored to specific types of graphs.
```

**Review by MyNameIsTrez — APPROVED** [review]
Really nice story and use of graphs! :D

**Comment by vsaulue** [comment]
Please replace `factorio-tech-tree-spacing.png` with this version (removed lot of empty space).

factorio-tech-tree-spacing
